File: GridView/FilterinIngIndicatorsInGridView/FilterinIngIndicatorsInGridViewVB/MyGridFilterCellElement.vb
Imports Telerik.WinControls
Imports Telerik.WinControls.UI

#Region "CustomGridFilterCellElement"
Public Class MyGridFilterCellElement
    Inherits GridFilterCellElement

    Private _clearButtonElement As RadButtonElement

    Public Sub New(ByVal column As GridViewDataColumn, ByVal row As GridRowElement)
        MyBase.New(column, row)
    End Sub

    Protected Overrides ReadOnly Property ThemeEffectiveType As Type
        Get
            Return GetType(GridFilterCellElement)
        End Get
    End Property

    Public ReadOnly Property ClearButtonElement As RadButtonElement
        Get
            Return Me._clearButtonElement
        End Get
    End Property

    Protected Overrides Sub CreateChildElements()
        MyBase.CreateChildElements()
        Me._clearButtonElement = New RadButtonElement()
        Me._clearButtonElement.Image = Image.FromFile("..\..\cross16x16.png")
        AddHandler Me._clearButtonElement.Click, AddressOf ClearButtonElement_Click
        Me.ClearButtonElement.NotifyParentOnMouseInput = False
        Me.Children.Add(Me.ClearButtonElement)
    End Sub

    Protected Overrides Function ArrangeOverride(ByVal finalSize As SizeF) As SizeF
        Dim size As SizeF = MyBase.ArrangeOverride(finalSize)

        If Me.IsFilterApplied Then
            Dim arrangeRect As RectangleF = New RectangleF(finalSize.Width - Me.FilterButton.DesiredSize.Width - Me.ClearButtonElement.DesiredSize.Width - Me.ElementSpacing, 4, Me.ClearButtonElement.DesiredSize.Width, Me.ClearButtonElement.DesiredSize.Height)
            Me.ClearButtonElement.Arrange(arrangeRect)
        End If

        If TypeOf Me.ColumnInfo Is GridViewDecimalColumn Then
            Me.TextAlignment = ContentAlignment.MiddleLeft
        End If

        Return size
    End Function

    Protected Overrides Sub SetTextAlignment()
        MyBase.SetTextAlignment()

        If TypeOf Me.ColumnInfo Is GridViewDecimalColumn Then
            Me.TextAlignment = ContentAlignment.MiddleLeft
        End If
    End Sub

    Protected Overrides Sub SetContentCore(ByVal value As Object)
        MyBase.SetContentCore(value)
        Me.ClearButtonElement.Visibility = If(Me.DataColumnInfo.FilterDescriptor IsNot Nothing, ElementVisibility.Visible, ElementVisibility.Collapsed)
    End Sub

    Public Overrides Function IsCompatible(ByVal data As GridViewColumn, ByVal context As Object) As Boolean
        Return TypeOf context Is GridFilterRowElement
    End Function

    Private Sub ClearButtonElement_Click(ByVal sender As Object, ByVal e As EventArgs)
        Me.DataColumnInfo.FilterDescriptor = Nothing
    End Sub
End Class
#End Region

File: GridView/FilterinIngIndicatorsInGridView/FilterinIngIndicatorsInGridView/MyGridFilterCellElement.cs
using System;
using System.Collections.Generic;
using System.Drawing;
using System.Linq;
using System.Text;
using System.Threading.Tasks;
using Telerik.WinControls;
using Telerik.WinControls.UI;

namespace FilterinIngIndicatorsInGridView
{
    #region CustomGridFilterCellElement
    public class MyGridFilterCellElement : GridFilterCellElement
    {
        private RadButtonElement clearButtonElement;

        public MyGridFilterCellElement(GridViewDataColumn column, GridRowElement row)
            : base(column, row)
        {
        }

        protected override Type ThemeEffectiveType
        {
            get
            {
                return typeof(GridFilterCellElement);
            }
        }

        public RadButtonElement ClearButtonElement
        {
            get
            {
                return this.clearButtonElement;
            }
        }

        protected override void CreateChildElements()
        {
            base.CreateChildElements();

            this.clearButtonElement = new RadButtonElement();
            this.clearButtonElement.Image = Image.FromFile(@"..\..\cross16x16.png");
            this.clearButtonElement.Click += ClearButtonElement_Click;
            this.clearButtonElement.NotifyParentOnMouseInput = false;
            this.Children.Add(this.clearButtonElement);
        }

        protected override SizeF ArrangeOverride(SizeF finalSize)
        {
            SizeF size = base.ArrangeOverride(finalSize);
            if (this.IsFilterApplied)
            {
                RectangleF arrangeRect = new RectangleF(finalSize.Width - this.FilterButton.DesiredSize.Width - this.clearButtonElement.DesiredSize.Width - this.ElementSpacing, 4,
                    this.clearButtonElement.DesiredSize.Width, this.clearButtonElement.DesiredSize.Height);
                this.clearButtonElement.Arrange(arrangeRect);
            }

            if (this.ColumnInfo is GridViewDecimalColumn)
            {
                this.TextAlignment = ContentAlignment.MiddleLeft;
            }

            return size;
        }

        protected override void SetTextAlignment()
        {
            base.SetTextAlignment();

            if (this.ColumnInfo is GridViewDecimalColumn)
            {
                this.TextAlignment = ContentAlignment.MiddleLeft;
            }
        }

        protected override void SetContentCore(object value)
        {
            base.SetContentCore(value);

            this.clearButtonElement.Visibility = this.DataColumnInfo.FilterDescriptor != null ? ElementVisibility.Visible : ElementVisibility.Collapsed;
        }

        public override bool IsCompatible(GridViewColumn data, object context)
        {
            return context is GridFilterRowElement;
        }

        private void ClearButtonElement_Click(object sender, EventArgs e)
        {
            this.DataColumnInfo.FilterDescriptor = null;
        }
    }

    #endregion
}
